### PR TITLE
Fix stream type to generic in comms default hooks

### PR DESCRIPTION
**Pull request by jeejakp12**
In comms default_hooks - decompress stream is hardcoded to cuda type. fix this to use generic type based on the grad tensor device.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @rohan-varma

### CODE REVIEW

**Comment by jeejakp12** [comment]
@fegin can you please help review the fix?

**Review by jgong5 — APPROVED** [review]

**Comment by jgong5** [comment]
@fegin does this change look good to you? Thanks!

**Comment by bsochack** [comment]
@fegin can we merge this change?

File: torch/distributed/algorithms/_comm_hooks/default_hooks.py [hunk 1]
```diff
@@ -73,8 +73,14 @@ def _decompress(state: LowPrecisionState, grad: torch.Tensor):
     """
     orig_grad_data = grad.data
     grad.data = grad.data.to(state.parameter_type)
+    try:
+        backend = getattr(torch, grad.device.type)
+    except AttributeError as e:
```
Comment (fegin) [inline torch/distributed/algorithms/_comm_hooks/default_hooks.py:78]: Just curious, when will this happen? If the backend is not registered then to the grad.device should not be that type?

**Review by fegin — APPROVED** [review]

File: torch/distributed/algorithms/_comm_hooks/default_hooks.py (hunk 1 above)
Comment (jgong5) [reply torch/distributed/algorithms/_comm_hooks/default_hooks.py:78]: It happens when pytorch works with a third-party device that is registered by out-of-the-tree extensions. The contract is that the device name is the same as the runtime namespace registered under `torch`, e.g., `torch.hpu` or `torch.xpu`.

@jeejakp12 Can you remind me if this works with the PrivateUse1 device?

Comment (jeejakp12) [reply torch/distributed/algorithms/_comm_hooks/default_hooks.py:78]: for PrivateUse1, we need to check the device type is PrivateUse1, then use torch._C._get_privateuse1_backend_name() to get the backend name. i will fix it to add the same.

Comment (jeejakp12) [reply torch/distributed/algorithms/_comm_hooks/default_hooks.py:82]: @jgong5 fixed the issue with PrivateUse1 device

Comment (jgong5) [reply torch/distributed/algorithms/_comm_hooks/default_hooks.py:82]: LGTM. Thanks for the fix.

**Pull request closed without merging**